exporter/exporterhelper/common.go

```diff
@@ -27,195 +26,213 @@ type baseRequestSender struct {
 	}
-
-	for _, op := range options {
-		op(be)
+	for _, op := range opts {
+		op(sets)
 	}
-	be.connectSenders()
 
-	// If retry sender is disabled then disable requeuing in the queue sender.
-	// TODO: Make re-enqueuing configurable on queue sender instead of relying on retry sender.
-	if qs, ok := be.queueSender.(*queueSender); ok {
-		// if it's not retrySender, then it is disabled.
-		if _, ok = be.retrySender.(*retrySender); !ok {
-			qs.requeuingEnabled = false
+	be := &baseExporter{
+		obsrep:          obsReport,
+		consumerOptions: sets.consumerOptions,
+	}
+	// Create senders in reverse order:
+	// Start with the basic exporter wrapper.
+	be.exporterSender = &exporterSender{StartFunc: sets.startFunc, ShutdownFunc: sets.shutdownFunc}
+	be.addSender(be.exporterSender)
+	// Before that add the timeout sender.
+	be.addSender(newTimeoutSender(sets.timeoutConfig, be.lastSender))
+	// Before that add the retry sender if configured.
+	if sets.retryConfig != nil {
+		if !sets.retryConfig.Enabled {
+			// TODO: Simplify logic for error logging request. Consider to add it always if both retry_on_failure is not enabled, not only when configured but disabled.
+			be.addSender(&errorLoggingRequestSender{
+				baseRequestSender: baseRequestSender{nextSender: be.lastSender},
+				logger:            set.Logger,
+				message:           "Exporting failed. Try enabling retry_on_failure config option to retry on retryable errors",
+			})
+		} else {
+			be.retrySender = newRetrySender(*sets.retryConfig, set, be.lastSender)
+			be.addSender(be.retrySender)
+		}
+	}
+	// Before that add the observability sender.
+	be.addSender(osf(obsReport, be.lastSender))
+	// Before that add the queue sender if configured.
+	if sets.queueConfig != nil {
+		if !sets.queueConfig.Enabled {
+			// TODO: Consider to remove this. Not sure how this helps with dropping data.
+			be.addSender(&errorLoggingRequestSender{
+				baseRequestSender: baseRequestSender{nextSender: be.lastSender},
+				logger:            set.Logger,
+				message:           "Exporting failed. Dropping data. Try enabling sending_queue to survive temporary failures.",
+			})
+		} else {
+			qs := newQueueSender(*sets.queueConfig, set, signal, marshaler, unmarshaler, be.lastSender)
+			// if it's not retrySender, then it is disabled.
+			if sets.retryConfig != nil && !sets.retryConfig.Enabled {
+				qs.requeuingEnabled = false
+			}
+			be.queueSender = qs
+			be.addSender(be.queueSender)
 		}
 	}
 
 	return be, nil
 }
 
-// send sends the request using the first sender in the chain.
-func (be *baseExporter) send(ctx context.Context, req Request) error {
-	return be.queueSender.send(ctx, req)
-}
-
-// connectSenders connects the senders in the predefined order.
-func (be *baseExporter) connectSenders() {
-	be.queueSender.setNextSender(be.obsrepSender)
-	be.obsrepSender.setNextSender(be.retrySender)
-	be.retrySender.setNextSender(be.timeoutSender)
+func (be *baseExporter) addSender(sender requestSender) {
```

Comment:
It'd be good to have a comment here that the sender is set in front of the current chain.